src/common/language.js

```diff
@@ -0,0 +1,35 @@
+/**
+ * Picks the most preferred of the available languages. Comparison only on
+ * the primary language tag. Region, variant, and script tags are ignored.
+ *
+ * @param {readonly string[]} availableLanguages: Languages to pick from.
+ * @param {readonly string[]} preferredLanguages: Languages to prefer.
+ * @return {string|undefined} matching language, if any.
+ * @example
+ *   matchLanguageTags(['en', 'ru', 'uk'], ['be-BY', 'ru-BY', 'en-UK']) // => 'ru'
+ */
+export function pickPreferredLanguage(availableLanguages, preferredLanguages) {
+  for (const preferredLanguage of preferredLanguages) {
+    const preferredLanguageCode = languageTagCode(preferredLanguage);
+    for (const availableLanguage of availableLanguages) {
+      const availableLanguageCode = languageTagCode(availableLanguage);
+      if (availableLanguageCode === preferredLanguageCode)
+        return availableLanguage;
+    }
+  }
+}
+
+/**
+ * Takes the language tag as per RFC5646 ("uk", "uk-UA", "ru-BY", en-GB").
+ * Returns the primary language-code subtag (lower-case, ISO 639).
+ *
+ * @param {string} languageTag language tag with one or more subtags
+ * @return {string|undefined} first subtag, two/three lowercase letters
+ * @example languageTagCode('en-US') // => 'en'
+ * @example languageTagCode('uk') // => 'uk'
+ * @see https://tools.ietf.org/html/rfc5646
+ */
+export function languageTagCode(languageTag) {
+  const matches = languageTag.toLowerCase().match(/^[a-z]{2,3}/);
+  return (matches ?? [])[0];
```

Comment:
When the `??` branch is taken, this evaluates to `undefined`. Is this expected behaviour?

Reply:
Yes. As indicated by `@returns {string|undefined}`. I could write it as `matches?.[0]`... Will look less dodgy.

But now that I think of it. The `languageTag` is not valid if it doesn't match that regexp. In that case it's most certainly a bug somewhere up at the call site. So it makes sense to throw `RangeError` instead of silently returning `undefined`.

Or, what would be the preferred error mode in your opinion? IMO if it silently fails like now then `pickLanguageTag` should check that and not use `undefined` as a valid match. But I am leaning towards throwing an error.